xwiki-platform-core/xwiki-platform-image/xwiki-platform-image-style/xwiki-platform-image-style-api/src/main/java/org/xwiki/image/style/internal/rendering/AbstractCaptionedImageChainingListener.java

```diff
@@ -0,0 +1,112 @@
+import org.xwiki.rendering.listener.reference.ResourceReference;
+
+/**
+ * Base class for chaining listeners that handle captioned images.
+ *
+ * @version $Id$
+ * @since 14.10.12
+ * @since 15.5RC1
+ */
+abstract class AbstractCaptionedImageChainingListener extends LookaheadChainingListener
+{
+    static final String DATA_XWIKI_IMAGE_STYLE_TEXT_WRAP = "data-xwiki-image-style-text-wrap";
+
+    static final String WIDTH_PROPERTY = "width";
+
+    static final String STYLE_PROPERTY = "style";
+
+    static final String STYLE_SEPARATOR = ";";
+
+    static final String DATA_XWIKI_IMAGE_STYLE = "data-xwiki-image-style";
+
+    static final String DATA_XWIKI_IMAGE_STYLE_ALIGNMENT = "data-xwiki-image-style-alignment";
+
+    static final String DATA_XWIKI_IMAGE_STYLE_BORDER = "data-xwiki-image-style-border";
+
+    protected final Deque<Map<String, String>> figureParametersQueue = new ArrayDeque<>();
+
+    protected AbstractCaptionedImageChainingListener(ListenerChain listenerChain)
+    {
+        super(listenerChain, 2);
+    }
+
+    @Override
+    public void beginFigure(Map<String, String> parameters)
+    {
+        this.figureParametersQueue.push(parameters);
+        super.beginFigure(parameters);
+    }
+
+    @Override
+    public void onImage(ResourceReference reference, boolean freestanding, String id, Map<String, String> parameters)
+    {
+        QueueListener.Event figureEvent = getPreviousFigureEvent();
+        if (figureEvent != null) {
+            // Merge the image parameters to the figure parameters when the image is wrapped in a figure.
+            Object[] eventParameters = figureEvent.eventParameters;
+            // Sanity check to make sure that we are handling the expected case.
+            if (eventParameters.length == 1 && eventParameters[0] instanceof Map<?, ?>) {
+                Map<String, String> figureParameters = this.figureParametersQueue.pop();
+                Map<String, String> updatedFigureParameters = updateFigureParameters(figureParameters, parameters);
+                this.figureParametersQueue.push(updatedFigureParameters);
+                eventParameters[0] = updatedFigureParameters;
+            }
+        }
+
+        super.onImage(reference, freestanding, id, parameters);
+    }
+
+    @Override
+    public void endFigure(Map<String, String> parameters)
+    {
+        super.endFigure(this.figureParametersQueue.pop());
+    }
+
+    protected abstract Map<String, String> updateFigureParameters(Map<String, String> figureParameters,
+        Map<String, String> imageParameters);
+
+    private QueueListener.Event getPreviousFigureEvent()
+    {
+        QueueListener.Event result = null;
+
+        QueueListener.Event candidate = getPreviousEvents().peekLast();
+        if (candidate != null && candidate.eventType == EventType.BEGIN_LINK) {
+            result = candidate;
+            candidate = getPreviousEvents().get(getPreviousEvents().size() - 2);
```

Comment:
This together with the increased lookahead (2 vs. 1) is the main change, the rest is refactoring.